core/lib/zksync_core/src/basic_witness_input_producer/mod.rs

```diff
@@ -0,0 +1,186 @@
+use zksync_types::{L1BatchNumber, L2ChainId};
+
+use async_trait::async_trait;
+use tokio::runtime::Handle;
+use tokio::task::JoinHandle;
+
+mod vm_interactions;
+
+use crate::basic_witness_input_producer::vm_interactions::start_next_miniblock;
+use vm_interactions::{create_vm, execute_tx};
+use zksync_types::block::MiniblockExecutionMode;
+
+/// Component that extracts all data (from DB) necessary to run a Basic Witness Generator.
+/// This component will upload Witness Inputs to the object store.
+/// This allows Witness Generator workflow (that needs only Basic Witness Generator Inputs)
+/// to be run only using the object store information, having no other external dependency.
+#[derive(Debug)]
+pub struct BasicWitnessInputProducer {
+    connection_pool: ConnectionPool,
+    validation_computational_gas_limit: u32,
+    l2_chain_id: L2ChainId,
+    object_store: Arc<dyn ObjectStore>,
+}
+
+impl BasicWitnessInputProducer {
+    pub async fn new(
+        connection_pool: ConnectionPool,
+        store_factory: &ObjectStoreFactory,
+        l2_chain_id: L2ChainId,
+    ) -> anyhow::Result<Self> {
+        let validation_computational_gas_limit = u32::MAX;
+        Ok(BasicWitnessInputProducer {
+            connection_pool,
+            validation_computational_gas_limit,
+            object_store: store_factory.create_store().await.into(),
+            l2_chain_id,
+        })
+    }
+
+    fn process_job_impl(
+        rt_handle: Handle,
+        l1_batch_number: L1BatchNumber,
+        started_at: Instant,
+        connection_pool: ConnectionPool,
+        validation_computational_gas_limit: u32,
+        l2_chain_id: L2ChainId,
+    ) -> anyhow::Result<WitnessBlockState> {
+        let mut connection = rt_handle.block_on(connection_pool.access_storage())?;
+
+        let miniblocks_execution_data = rt_handle.block_on(
+            connection
+                .transactions_dal()
+                .get_miniblocks_to_execute_for(MiniblockExecutionMode::L1Batch(l1_batch_number)),
+        );
+
+        let (mut vm, storage_view) = create_vm(
+            rt_handle.clone(),
+            l1_batch_number,
+            connection,
+            validation_computational_gas_limit,
+            l2_chain_id,
+        );
+
+        tracing::info!("Started execution of l1_batch: {l1_batch_number:?}");
+        for (index, miniblock_execution_data) in miniblocks_execution_data.iter().enumerate() {
+            tracing::debug!(
+                "Started execution of miniblock: {:?}",
+                miniblock_execution_data.number
+            );
+            for tx in &miniblock_execution_data.txs {
+                tracing::debug!("Started execution of tx: {tx:?}");
+                execute_tx(tx, &mut vm);
+                tracing::debug!("Finished execution of tx: {tx:?}");
+            }
+            let next_miniblock_execution_data = if index + 1 < miniblocks_execution_data.len() {
+                Some(&miniblocks_execution_data[index + 1])
```

Comment:
you can just finish the loop here and do not set the new miniblock.

Reply:
yes, i understand it. 
But when you return None, you call the function, which do literally noting if it's none. 
So for some you can call this function and for none just do nothing and remove some clunky code from set_new_miniblock